### PR TITLE
Notifications: small fixes found after reviewer

### PR DESCRIPTION
**Pull request by humitos**
Related to https://github.com/readthedocs/readthedocs.org/pull/10922#discussion_r1440756414.

Screenshot_2024-01-04_16-30-05

### CODE REVIEW

File: readthedocs/notifications/messages.py [hunk 1]
```diff
@@ -29,8 +34,18 @@ def __repr__(self):
...
 
+    def _escape_format_values(self, format_values):
+        """
+        Escape all potential HTML tags included in format values.
+
+        This is a protection against rendering potential values defined by the user.
+        It uses the Django's util function ``escape`` (similar to ``|escape`` template tag filter)
+        to convert HTML characters into regular characters.
+        """
+        return {key: escape(value) for key, value in format_values.items()}
```
Comment (stsewd) [inline readthedocs/notifications/messages.py:45]: This means that nested values won't be supported. I think this should be fine, but probably worth documenting somewhere if it isn't already to avoid surprises.

File: readthedocs/notifications/messages.py [hunk 2]
```diff
@@ -55,10 +70,20 @@ def get_display_icon_classes(self):
         return " ".join(classes)
 
     def get_rendered_header(self):
-        return self.header.format(**self.format_values)
+        try:
+            return self.header.format(**self.format_values)
+        except KeyError:
+            # There was a key missing
+            log.exception("There was a missing key when formating a header's Message.")
+            return self.header.format_map(defaultdict(str, **self.format_values))
```
Comment (stsewd) [inline readthedocs/notifications/messages.py:78]: Since we have control over the notification, I think this should fail hard, our tests should catch these instead of silently failing.

Comment (humitos) [reply readthedocs/notifications/messages.py:78]: We can't catch this by tests since this rendering happens at runtime. This could happen because a `Notification` was created with invalid format values, or because we have change the message over time and started required more format values.

File: readthedocs/notifications/messages.py (hunk 1 above)
Comment (humitos) [reply readthedocs/notifications/messages.py:45]: This is a good point. Currently, we don't need/support nested values so I think it's fine to write down a comment here for now 👍🏼

Comment (humitos) [reply readthedocs/notifications/messages.py]: Done in 24bd75d92

**Comment by humitos** [comment]
@stsewd I'd like to get an approve here so we can deploy this tomorrow. Can you take a final look at it? 🙏🏼

File: readthedocs/notifications/messages.py [hunk 3]
```diff
@@ -29,8 +34,26 @@ def __repr__(self):
...
+        to convert HTML characters into regular characters.
+
+        NOTE: currently, we don't support values that are not ``str`` or ``int``.
+        If we want to support other types or nested dictionaries,
+        we will need to iterate recursively to apply the ``escape`` function.
+        """
+        return {
+            key: escape(value)
+            for key, value in format_values.items()
+            if isinstance(value, (str, int))
```
Comment (stsewd) [inline readthedocs/notifications/messages.py:52]: I think this is fine without this check, there are more built-in types that can be included (like bool).

```suggestion
```

We should have more tests instead.

File: readthedocs/notifications/messages.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/notifications/messages.py:86]: The idea here would be to test for each notification we generate on each part of the code base. If we generate a notification when the user does something, we should have a test for it.

**Review by stsewd — APPROVED** [review]

File: readthedocs/notifications/messages.py (hunk 3 above)
Comment (humitos) [reply readthedocs/notifications/messages.py:52]: Yeah, we can expand to other types in the future. I want to avoid running `escape(dict)` and returning 500. For now, `str` and `int` are the ones we are using.

File: readthedocs/notifications/messages.py (hunk 2 above)
Comment (humitos) [reply readthedocs/notifications/messages.py:86]: That's fine, we can test _the current state_, but it could fail still in the future:

1. we have a notification with `body="Go to {url} and put {code}."` and `format_values={"url": "https://..", "code": 1234}`
2. in the future, we change the message to `body="Go to {url} and put {code}. Otherwise, contact support at {email}`

At this point, all the old notifications with only `url` and `code` as format values will start to fail because they are missing the `email` key. This check here is to protect ourselves against this issue.

Comment (humitos) [reply readthedocs/notifications/messages.py:86]: We will find some edge cases as we go that we can improve and write test cases for them and protections like this one against them.
